src/compiler/checker.ts

```diff
@@ -18870,6 +18870,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const singleton = elementFlags[i] & ElementFlags.Variadic ? t :
                 elementFlags[i] & ElementFlags.Rest ? createArrayType(t) :
                 createTupleType([t], [elementFlags[i]]);
+            // avoid infinite recursion, if the singleton is the type variable itself
+            // then we'd just get back here with the same arguments from within instantiateMappedType
+            if (singleton === typeVariable) {
+                return mappedType;
+            }

```

Comment:
I don't really expect this to be the final solution to the problem. I think that it's likely to be possible to craft an extra test case with such an infinite recursion that would involve more than one layer between `singleton` and `typeVariable`. I wasn't yet able to manufacture such a test case though.

I'm not sure what's the best strategy to avoid such recursions in the compiler right now - there are different mechanisms like that in the codebase but I wasn't sure which one should be used here.

I imagine that we'd have to track what mapped types are being instantiated, and with what type arguments. This would allow us to bail out at some level. Unless perhaps - this is just a manifestation of the issue that lies elsewhere.

The other test case (with a type alias) works OK because it's able to remap that `typeVariable` to the alias' own type parameter~ since that alias has `.mapper` property set. So the `mapper` available here is the one that was combined [here](https://github.dev/microsoft/TypeScript/blob/f64f40d20511d2d2a9eb4fcc1ad6750977be2d40/src/compiler/checker.ts#L18714-L18717). That is not the case for an inline mapped type though since that has no `.mapper`